src/proteus/dswx_hls.py

```diff
@@ -1744,21 +1747,66 @@ def _compute_and_apply_cloud_layer(wtr_2_layer, fmask,
         logger.info(error_msg)
         raise Exception(error_msg)
 
-    logger.info(f'mask adjacent to cloud/cloud-shadow mode:'
-                f' {mask_adjacent_to_cloud_mode}')
- 
     # Check Fmask cloud shadow bit (3) => bit 0
-    cloud_layer[np.bitwise_and(fmask, 2**3) == 2**3] = 1
+    preliminary_cloud_layer[np.bitwise_and(fmask, 2**3) == 2**3] = 1
 
     if mask_adjacent_to_cloud_mode == 'mask':
         # Check Fmask adjacent to cloud/shadow bit (2) => bit 0
-       cloud_layer[np.bitwise_and(fmask, 2**2) == 2**2] = 1
+       preliminary_cloud_layer[np.bitwise_and(fmask, 2**2) == 2**2] = 1
 
     # Check Fmask cloud bit (1) => bit 2
-    cloud_layer[np.bitwise_and(fmask, 2**1) == 2**1] += 4
+    preliminary_cloud_layer[np.bitwise_and(fmask, 2**1) == 2**1] += 4
 
-    # If cloud (1) or cloud shadow (3), mark WTR as WTR_CLOUD_MASKED
-    wtr_layer[cloud_layer != 0] = WTR_CLOUD_MASKED
+    return preliminary_cloud_layer
+
+
+
+def _add_snow_to_cloud_layer(wtr_2_layer, cloud_layer, fmask,
```

Comment:
Currently, this docstring says that the function takes in an argument named `cloud_layer`, and then returns an argument named `cloud_layer`. It's not great practice to use the same variable name for inputs and outputs. Plus, in the code itself, it looks like the input `cloud_layer` is being modified directly and so there's no need for it to be returned.

Two suggestions for how to approach this:
1) change the input variable name to be `preliminary_cloud_layer`, then make a copy of it in the code and return `cloud_layer`. This is the style of other functions in the codebase, and much clearer about what the input should be.
2) update the docstring to say that the input cloud layer is modified by the function, make sure to detail _how_ the function is modifying the actual values of the input layer, and also remove the `Return` statement. This strategy is less preferable to me, as it is contrary to the coding style used elsewhere in the code.

Reply:
I prefer to avoid making an unnecessary copies of the array. I think the code "as is" is easier to follow.

Reply:
In that case, could you please follow suggestion 2?

Reply:
I think the way it is now it's easier to understand. `cloud_layer` is mutable, but by returning it, we are making explicit that that's the output of the function. The input `preliminary_cloud_layer` and output `cloud_layer` are the same array, but I think the code this way is easier to follow. I updated the docstrings to make this point more clear:
```       cloud_layer : numpy.ndarray
            Preliminary cloud mask (without the snow/ice class). This array
            is updated within the function (mutable numpy.ndarray) and returned
            after the snow/ice class is added to the layer.```

Reply:
I agree that it's better to be explicit, both for the input arguments and the output arguments. In that case, please go with suggestion 1. There are so many other places in the code where extra copies of arrays are being created, I don't think that one additional occurance of that is going to matter.

Reply:
I don't see a reason to add an unnecessary copy of data. The other array copies in the code are needed to create different layers. If that's not the case, please let me know because then we should fix that too.

Reply:
Is there a third approach? The requirements for the new approach would be that this function can not both mutate an input array and return that same array. It must only do one of these things. This is significantly cleaner.

Some pros and cons of each option: Mutating an input array makes for code that is very hard to debug and understand (as an example, mutating input arguments was a significant reason why the original QA code was nearly impossible to understand), but it avoids copying an array. Copying the array is much cleaner for the user to understand and is one of the foundations of functional programming, but does create a (temporary) copy of the array.

Reply:
I guess the third approach is what we are doing lol. The behavior is described in the docstrings and also, this function intended for internal use (startswith `_`), so it won't be used by other modules. It's also very specific to a part of the code (add snow class to the cloud layer). Finally, having `preliminary_cloud_layer` and `cloud_layer` make it clear what is preliminary and final.